service/worker/scanner/history/scavenger_test.go

```diff
@@ -466,3 +520,102 @@ func (s *ScavengerTestSuite) TestMixesTwoPages() {
+				Info:        persistence.BuildHistoryGarbageCleanupInfo("namespaceID2", "workflowID2", "runID2"),
+			},
+		},
+	}, nil)
+
+	db.EXPECT().GetAllHistoryTreeBranches(gomock.Any(), &persistence.GetAllHistoryTreeBranchesRequest{
+		PageSize:      pageSize,
+		NextPageToken: []byte("page1"),
+	}).Return(&persistence.GetAllHistoryTreeBranchesResponse{
+		Branches: []persistence.HistoryBranchDetail{
+			{
+				BranchToken: s.toBranchToken("treeID3", "branchID3"),
+				ForkTime:    timestamp.TimeNowPtrUtcAddDuration(-scvgr.historyDataMinAge() * 2),
+				Info:        persistence.BuildHistoryGarbageCleanupInfo("namespaceID3", "workflowID3", "runID3"),
+			},
+			{
+				BranchToken: s.toBranchToken("treeID4", "branchID4"),
+				ForkTime:    timestamp.TimeNowPtrUtcAddDuration(-scvgr.historyDataMinAge() * 2),
+				Info:        persistence.BuildHistoryGarbageCleanupInfo("namespaceID4", "workflowID4", "runID4"),
+			},
+		},
+	}, nil)
+	mockedNamespace := namespace.NewNamespaceForTest(
+		nil,
+		&persistencepb.NamespaceConfig{Retention: timestamp.DurationPtr(time.Hour)},
+		false,
+		nil,
+		0,
+	)
+	workflowInRetention := &historyservice.DescribeMutableStateResponse{
+		DatabaseMutableState: &persistencepb.WorkflowMutableState{
+			ExecutionInfo: &persistencepb.WorkflowExecutionInfo{
+				LastUpdateTime: timestamp.TimePtr(time.Now()),
+			},
+		},
+	}
+	workflowPassedRetention := &historyservice.DescribeMutableStateResponse{
+		DatabaseMutableState: &persistencepb.WorkflowMutableState{
+			ExecutionInfo: &persistencepb.WorkflowExecutionInfo{
+				LastUpdateTime: timestamp.TimePtr(time.Now().Truncate(time.Hour * 3)),
+			},
+		},
+	}
+	registry.EXPECT().GetNamespaceByID(gomock.Any()).Return(mockedNamespace, nil).AnyTimes()
+	client.EXPECT().DescribeMutableState(gomock.Any(), &historyservice.DescribeMutableStateRequest{
+		NamespaceId: "namespaceID1",
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: "workflowID1",
+			RunId:      "runID1",
+		},
+	}).Return(workflowInRetention, nil)
+	client.EXPECT().DescribeMutableState(gomock.Any(), &historyservice.DescribeMutableStateRequest{
+		NamespaceId: "namespaceID2",
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: "workflowID2",
+			RunId:      "runID2",
+		},
+	}).Return(workflowPassedRetention, nil)
+	client.EXPECT().DescribeMutableState(gomock.Any(), &historyservice.DescribeMutableStateRequest{
+		NamespaceId: "namespaceID3",
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: "workflowID3",
+			RunId:      "runID3",
+		},
+	}).Return(workflowInRetention, nil)
+	client.EXPECT().DescribeMutableState(gomock.Any(), &historyservice.DescribeMutableStateRequest{
+		NamespaceId: "namespaceID4",
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: "workflowID4",
+			RunId:      "runID4",
+		},
+	}).Return(workflowPassedRetention, nil)
+	adminClient.EXPECT().DeleteWorkflowExecution(gomock.Any(), gomock.Any()).Return(nil, nil).Times(2)
```

Comment:
I think we should verify which workflows were actually deleted here. You could do this by using .Do instead of .Return, adding the requests to a list, and then verifying that list after the scavenger is run.

Reply:
Verify workflow id and run id